url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/StandardBeanInfo.java

```diff
@@ -0,0 +1,1505 @@
+// the License.  You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+//
+
+package org.msgpack.template.builder.beans;
+
+import static org.msgpack.template.builder.beans.Introspector.decapitalize;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.EventListener;
+import java.util.EventObject;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TooManyListenersException;
+
+class StandardBeanInfo extends SimpleBeanInfo {
+
+    // Prefixes for methods that set or get a Property
+    private static final String PREFIX_IS = "is"; //$NON-NLS-1$
+
+    private static final String PREFIX_GET = "get"; //$NON-NLS-1$
+
+    private static final String PREFIX_SET = "set"; //$NON-NLS-1$
+
+    // Prefix and suffix for Event related methods
+    private static final String PREFIX_ADD = "add"; //$NON-NLS-1$
+
+    private static final String PREFIX_REMOVE = "remove"; //$NON-NLS-1$
+
+    private static final String SUFFIX_LISTEN = "Listener"; //$NON-NLS-1$
+
+    private static final String STR_NORMAL = "normal"; //$NON-NLS-1$
+
+    private static final String STR_INDEXED = "indexed"; //$NON-NLS-1$
+
+    private static final String STR_VALID = "valid"; //$NON-NLS-1$
+
+    private static final String STR_INVALID = "invalid"; //$NON-NLS-1$
+
+    private static final String STR_PROPERTY_TYPE = "PropertyType"; //$NON-NLS-1$
+
+    private static final String STR_IS_CONSTRAINED = "isConstrained"; //$NON-NLS-1$
+
+    private static final String STR_SETTERS = "setters"; //$NON-NLS-1$
+
+    private static final String STR_GETTERS = "getters"; //$NON-NLS-1$
+
+    private boolean explicitMethods = false;
+
+    private boolean explicitProperties = false;
+
+    private boolean explicitEvents = false;
+
+    private BeanInfo explicitBeanInfo = null;
+
+    private EventSetDescriptor[] events = null;
+
+    private MethodDescriptor[] methods = null;
+
+    private PropertyDescriptor[] properties = null;
+
+    private BeanDescriptor beanDescriptor = null;
+
+    BeanInfo[] additionalBeanInfo = null;
+
+    private Class<?> beanClass;
+
+    private int defaultEventIndex = -1;
+
+    private int defaultPropertyIndex = -1;
+
+    private static PropertyComparator comparator = new PropertyComparator();
+
+    private boolean canAddPropertyChangeListener;
+
+    private boolean canRemovePropertyChangeListener;
+
+    StandardBeanInfo(Class<?> beanClass, BeanInfo explicitBeanInfo, Class<?> stopClass)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 139-149 into a separate method.